src/main/scala/coreplex/Coreplex.scala

```diff
@@ -24,88 +26,74 @@ case object BankIdLSB extends Field[Int]
 case object BuildL2CoherenceManager extends Field[(Int, Parameters) => CoherenceAgent]
 /** Function for building some kind of tile connected to a reset signal */
 case object BuildTiles extends Field[Seq[(Bool, Parameters) => Tile]]
-/** A string describing on-chip devices, readable by target software */
-case object ConfigString extends Field[Array[Byte]]
-/** Number of external interrupt sources */
-case object NExtInterrupts extends Field[Int]
-/** Interrupt controller configuration */
-case object PLICKey extends Field[PLICConfig]
 /** The file to read the BootROM contents from */
 case object BootROMFile extends Field[String]
-/** Export an external MMIO slave port */
-case object ExportMMIOPort extends Field[Boolean]
-/** Expose additional TileLink client ports */
-case object NExternalClients extends Field[Int]
-/** Extra top-level ports exported from the coreplex */
-case object ExtraCoreplexPorts extends Field[Parameters => Bundle]
 
 trait HasCoreplexParameters {
   implicit val p: Parameters
-  lazy val nTiles = p(NTiles)
   lazy val nMemChannels = p(NMemoryChannels)
   lazy val nBanksPerMemChannel = p(NBanksPerMemoryChannel)
   lazy val nBanks = nMemChannels*nBanksPerMemChannel
   lazy val lsb = p(BankIdLSB)
   lazy val innerParams = p.alterPartial({ case TLId => "L1toL2" })
   lazy val outermostParams = p.alterPartial({ case TLId => "Outermost" })
   lazy val outermostMMIOParams = p.alterPartial({ case TLId => "MMIO_Outermost" })
-  lazy val nExtClients = p(NExternalClients)
-  lazy val exportMMIO = p(ExportMMIOPort)
+  lazy val configString = p(rocketchip.ConfigString).get
+  lazy val globalAddrMap = p(rocketchip.GlobalAddrMap).get
 }
 
-abstract class Coreplex(implicit val p: Parameters) extends Module
+case class CoreplexConfig(
+    nTiles: Int,
+    nExtInterrupts: Int,
+    nSlaves: Int,
+    hasSupervisor: Boolean,
+    hasExtMMIOPort: Boolean)
+{
+  val plicKey = PLICConfig(nTiles, hasSupervisor, nExtInterrupts, 0)
+}
+
+abstract class Coreplex(implicit val p: Parameters, implicit val c: CoreplexConfig) extends Module
     with HasCoreplexParameters {
-  class CoreplexIO(implicit val p: Parameters) extends Bundle {
-    val mem  = Vec(nMemChannels, new ClientUncachedTileLinkIO()(outermostParams))
-    val ext_clients = Vec(nExtClients, new ClientUncachedTileLinkIO()(innerParams)).flip
-    val mmio = p(ExportMMIOPort).option(new ClientUncachedTileLinkIO()(outermostMMIOParams))
-    val interrupts = Vec(p(NExtInterrupts), Bool()).asInput
+  class CoreplexIO(implicit val p: Parameters, implicit val c: CoreplexConfig) extends Bundle {
+    val master = new Bundle {
+      val mem = Vec(nMemChannels, new ClientUncachedTileLinkIO()(outermostParams))
```

Comment:
Yes, you're on the right track, and that should be the next thing that gets refactored.  I wanted to maintain the same functionality so that all the things that depended on it still worked.